Add implementation for Members, Sources and Invitations endpoints

GG-Yanne opened the pull request:
## Add support for managing users

This PR implements all endpoints required for managing users on an account as well as adding support for invitations and sources handling

agateau-gg commented on tests/test_client.py:
```diff
@@ -1391,3 +1396,71 @@ def test_read_metadata_remediation_message(client: GGClient):
+
+    result = client.list_members()
+
+    assert isinstance(result, CursorPaginatedResponse), result.content
```
It's better to use `result` as the error message: because `Detail` does not have a `content` field.

GG-Yanne commented on pygitguardian/client.py:
```diff
@@ -859,3 +898,84 @@ def scan_diff(
+
+        # We bypass `is_ok` because the response content type is none
+        if response.status_code == 204:
+            return 204
```
I changed all delete methods to return `Optional[Detail]` 👍

GG-Yanne commented on pygitguardian/models.py:
```diff
@@ -1077,3 +1094,146 @@ def __repr__(self) -> str:
+    active: bool
+
+
+class MemberSchema(BaseSchema):
```
I added a docstring explaining why dataclasses did not work natively, but perhaps we can go with a simple type alias for access levels and such, let me know what you think 🙌

GG-Yanne commented on pygitguardian/models.py:
```diff
@@ -965,25 +982,70 @@ class Feedback(Base, FromDictMixin):
-    source_criticality: str
-    last_scan: Optional[Dict[str, Any]]  # TODO: add LastScan
+    source_criticality: SourceCriticality
+    last_scan: Scan
```
It should be Optional too, adding it back 👍

GG-Yanne commented on tests/test_client.py:
```diff
@@ -1391,3 +1396,71 @@ def test_read_metadata_remediation_message(client: GGClient):
+]
+
+
+@my_vcr.use_cassette("test_list_members.yaml", ignore_localhost=False)
```
I am unsure why you think it would not work ? The `GITGUARDIAN_API_KEY` will require the related scopes (`members`, `teams` and `sources`) but it will work if those scopes are granted 🤔

GG-Yanne replied:
I gave a shot at writing tests that do not require mocking calls and rely only on cassettes, while feasible, it requires having few members and invitations sent in the workspace (also it will delete some resources so better do it on a test workspace), those are documented through comments on those tests. Let me know if you think it is fine the way it is now, I don't think mock would do great justice for the tests as you said :/

Alustrat-gg commented on pygitguardian/client.py:
```diff
@@ -1213,3 +1217,64 @@ def delete_team_member(
+        obj.status_code
+        return obj
+
+    def list_teams_sources(
```
list_team_sources

Alustrat-gg commented on pygitguardian/client.py:
```diff
@@ -1278,3 +1282,61 @@ def update_team_source(
+        obj.status_code
+        return obj
+
+    def send_invitation(
```
create_invitation

Alustrat-gg reviewed (approved):
Good Job Yanne 👍 

I just think that we sometime have incoherences in the way we pass parameters/queries/bodies but it's a detail

GG-Yanne commented:
> ## Regarding VCR cassettes
> If it's not possible to write a script to automatically setup the account, then can you create a Markdown document in doc/dev/ (We have a similar directory in ggshield, but you would have to create it there) to list all the requirements for the account in one place? This would be more efficient than having to look for comments in different tests to gather them all.

I added `doc/dev/tests.md` akin to what is done for GG Shield and listed all hard requirements for the workplace to be able to run the test suite without issues

> ### `update_*()` methods
> * either make them all return `Optional[Detail]` or `Union[Detail, <ModelClass>]` (update_team_source() returns `Union[Detail, int]`).

`update_team_source` works in a particular way, it does not return a ModelClass since we are adding and removing sources and the endpoint does not return, see : https://api.gitguardian.com/docs#tag/Team-Sources/operation/update-team-sources

In any case, I applied the same logic as for delete endpoints and optionally returning a Detail object if the status code is not the one we expect for a successful call 👍 

> * be consistent with argument names because people are going to use keyword arguments for them and changing them after a release is an API breakage. So either we call them all `payload` or w call theom all `<name_of_the_model>`. I prefer `payload` because what they take is not an instance of the model, it's an instance of `Update<Model>`.

I changed all updates to payload so they are all aligned 👍 

> ### Misc
> Rename `list_teams_sources()` to `list_team_sources()` for consistency.

This was already taken care of in a later commit 🙌 

> Remove the "obj.status_code" useless lines (`grep -n "^ *obj\.status_code$" client.py` to find them)

`grep -n "^ *obj\.status_code$" client.py` does not return anything anymore 👍 

> ## models.py
> ### `@post_load` methods
> * Type-hint for `**kwargs` must be `Any`.

I changed kwargs typing to be Any 👍 

> * All existing `@post_load` methods are called `make_<something>`. Can you use the same naming convention for yours?

I changed all @post_load method to be respect this convention

> ## conftest.py
> get_member() is never used, can we remove it?

Indeed, I removed it 👍 

> Instead of creating fixtures, you can create a test.utils module with plain functions and import them in the tests.

I create the corresponding utils file and updated references for these functions

agateau-gg reviewed (approved):
Looks good now! Thanks for this significant addition to the API and for documenting the test setup requirements 👍🏻.
